src/buffer/line_cache.rs

```diff
@@ -0,0 +1,115 @@
+        let mut line = 0;
+        loop {
+            let text = document.read_forward(offset);
+            if text.is_empty() { return; }
+            
+            let mut off = 0;
+            loop {
+                off = memchr2(b'\n', b'\n', text, off);
+                if off == text.len() { break; }
+
+                if line % CACHE_EVERY == 0 {
+                    self.cache.push((offset+off, line));
+                }
+                line += 1;
+                off += 1;
+            }
+
+            offset += text.len();
+        }
+    }
+
+    /// Updates the cache after a deletion.
+    /// `range` is the deleted byte range, and `text` is the content that was deleted.
+    pub fn delete(&mut self, range: Range<usize>, text: &Vec<u8>) {
+        let mut newlines = 0;
+        for c in text {
+            if *c == b'\n' {
+                newlines += 1;
+            }
+        }
+
+        if self.cache.len() == 0 { return; }
+        
+        let len = text.len();
+        let mut i = self.cache.len() - 1;
+        loop {
+            let (ref mut off, ref mut line) = self.cache[i];
+            if *off > range.start {
+                if *off < range.end {
+                    self.cache.remove(i); // cache point is within the deleted range
+                } else {
+                    *off -= len;
+                    *line -= newlines;
+                }
+            }
+            if i == 0 {
+                break;
+            }
+            i -= 1;
+        }
+    }
+
+    /// Updates the cache after an insertion.
+    /// `offset` is where the insertion occurs, and `text` is the inserted content.
+    pub fn insert(&mut self, offset: usize, text: &[u8]) {
+        // Count how many newlines were inserted
+        let mut newlines = 0;
+        for c in text {
+            if *c == b'\n' {
+                newlines += 1;
+            }
+        }
+
+        let len = text.len();
+        for &mut (ref mut off, ref mut line) in &mut self.cache {
+            if *off > offset {
+                *off += len;
+                *line += newlines;
+            }
+        }
+    }
+
+    /// Finds the nearest cached line-offset pair relative to a target line.
+    /// If `reverse` is false, it returns the closest *before* the target.
+    /// If `reverse` is true, it returns the closest *after or at* the target.
+    pub fn nearest_offset(&self, target_count: usize, reverse: bool) -> Option<(usize, usize)> {
+        let len = self.cache.len();
+        let mut i = 0;
+        while i < len {
+            if self.cache[i].1 >= target_count {
+                let ind = if !reverse {
+                    if i == 0 { return None; } // No previous line exists
+                    i - 1
+                } else {
+                    if i == len - 1 { return None; } // No next line exists
+                    i
+                };
+                return Some(self.cache[ind]);
+            }
+            i += 1;
```

Comment:
I believe this should ideally be a binary search, right?
If we use deltas as I proposed above, we could note the last deletion/insertion/nearest_offset position and its absolute index and cache it in a member. Then we can start iterating from there. Modifications and searches should occur near each other, after all.